Library/Homebrew/cmd/autoremove.rb

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+require "formula"
+require "cli/parser"
+
+module Homebrew
+  module_function
+
+  def autoremove_args
+    Homebrew::CLI::Parser.new do
+      usage_banner <<~EOS
+        `autoremove` [<options>]
+
+        Remove packages that weren't installed on request and are no longer needed.
+      EOS
+      switch "-n", "--dry-run",
+             description: "Just print what would be removed."
+      switch "-f", "--force",
+             description: "Remove without confirmation."
+      named 0
+    end
+  end
+
+  def get_removable_formulae(installed_formulae)
+    removable_formulae = []
+
+    installed_formulae.each do |formula|
+      # Reject formulae installed on request.
+      next if formula.installed_kegs.any? { |keg| Tab.for_keg(keg).installed_on_request }
+      # Reject formulae which are needed at runtime by other formulae.
+      next if installed_formulae.map(&:deps).flatten.uniq.map(&:to_formula).include?(formula)
+
+      removable_formulae << installed_formulae.delete(formula)
+      removable_formulae += get_removable_formulae(installed_formulae)
+    end
+
+    removable_formulae
+  end
+
+  def autoremove
+    args = autoremove_args.parse
+
+    removable_formulae = get_removable_formulae(Formula.installed)
+
+    return if removable_formulae.empty?
+
+    formulae_names = removable_formulae.map(&:full_name)
+
+    oh1 "Formulae that could be removed: " \
+      "#{formulae_names.sort.map(&Formatter.method(:identifier)).to_sentence}",
+        truncate: false
+
+    return if args.dry_run?
+
+    unless args.force?
+      ohai "Proceed?", <<~EOF
+        Enter:  yes
+        CTRL-C: no
+
+        To mark formulae as not removable run:
+            brew install PACKAGE
+      EOF
+      readline
+    end
+
+    system HOMEBREW_BREW_FILE, "rm", *formulae_names
```

Comment:
Could this call to the `uninstall` code rather than shelling out?

Reply:
Not really sure how to achieve that, I looked at what `cmd/uninstall.rb` does and had a hard time figuring what I should call in `cmd/autoremove.rb`. Any pointers are welcomed :)

Reply:
I think `cmd/uninstall.rb` will need to be refactored a little to separate the argument parsing from the uninstall logic. There's also some output to the console that might need to be handled differently outside of the "uninstall" command.

Reply:
Yes, agreed. Pretty much everything except argument parsing could be pulled into a new `install.rb` which should be called here.

Reply:
An `install.rb` already exists, which is a *Helper module for performing (pre-)install checks*. Should it go in there? Or would it make more sense to create a new `uninstall.rb` and put it in there?